clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java

```diff
@@ -911,11 +911,25 @@ private void throwIfProducerClosed() {
             // producer callback will make sure to call both 'callback' and interceptor callback
             Callback interceptCallback = new InterceptorCallback<>(callback, this.interceptors, tp);
 
+            RecordAccumulator.RecordAppendResult result = accumulator.append(tp, timestamp, serializedKey,
+                    serializedValue, headers, interceptCallback, remainingWaitMs, true);
+            
+            if (result.retryForNewBatch) {
+                partitioner.batchCompleted(record.topic(), cluster, partition);
+                partition = partition(record, serializedKey, serializedValue, cluster);
+                tp = new TopicPartition(record.topic(), partition);
+                
+                log.trace("Sending record {} with callback {} to topic {} partition {}", record, callback, record.topic(), partition);
```

Comment:
record.toString could be expensive, so it would be best to surround this in an if statement, like
```
if (log.isTraceEnabled()) {
  log.trace("Sending record {} with callback {} to topic {} partition {}", record, callback, record.topic(), partition);
}
```
Of course, it's a bit unclear that we should be logging `record.toString` at all here, since we don't log it in the other cases.  it would probably be enough to just log (at trace level) that we are retrying the partitioning for the particular topic because there was a new batch.

Reply:
I was repeating the log, as this exact log message is used above. I could modify it to say that we are retrying with a given partition due to the new batch.

Reply:
Yes, I think that's a good idea.  We would want to know that we are retrying due to the current batch.

We should also be surrounding this with an if statement-- the fact that it isn't right now is a (small) performance bug, which we don't want to perpetuate :)  We don't know how expensive `record.toString` is but we can guess it's not free.  Let's fix the call above too

Reply:
Seems like I accidentally removed it. Maybe back when I was having problems with Npath. Fixed now.